build/generate-licenses.go

```diff
@@ -115,8 +126,147 @@ func main() {
 			log.Fatalf("Failed to write new file. %s", err)
 		} else {
 			fmt.Printf("Written %s\n", out.Name())
+
+			// some license files have same content, so we need to detect these files and create a convert map into a file
+			// In InitClassifier, we will use this convert map to avoid adding same license content with different license name
+			md5, err := getSameFileMD5(preFile, out)
+			if err != nil {
+				log.Fatalf("Failed to get same file md5. %s", err)
+				continue
+			}
+			if md5 != "" {
+				_, ok := sameFiles[md5]
+				if !ok {
+					sameFiles[md5] = make([]string, 0)
+				}
+				if !slices.Contains(sameFiles[md5], preLicenseName) {
+					sameFiles[md5] = append(sameFiles[md5], preLicenseName)
+				}
+				sameFiles[md5] = append(sameFiles[md5], licenseName)
+			}
+			preFile = out
+			preLicenseName = licenseName
+		}
+	}
+
+	// generate convert license name map
+	sameLicenses := make(map[string]string)
+	for _, fileNames := range sameFiles {
+		key := getLicenseKey(fileNames)
+		for _, fileName := range fileNames {
+			sameLicenses[fileName] = key
 		}
 	}
+	// save convert license name map to file
+	bytes, err := json.Marshal(sameLicenses)
+	if err != nil {
+		log.Fatalf("Failed to create json bytes. %s", err)
+		return
+	}
+	out, err := os.Create(sameLicensesDestination)
+	if err != nil {
+		log.Fatalf("Failed to create new file. %s", err)
+	}
+	defer out.Close()
+	_, err = out.Write(bytes)
+	if err != nil {
+		log.Fatalf("Failed to write same licenses json file. %s", err)
+	}
 
 	fmt.Println("Done")
 }
+
+// getSameFileMD5 returns md5 of the input file, if the content of input files are same
+func getSameFileMD5(f1, f2 *os.File) (string, error) {
```

Comment:
When reading through this function signature, I have no idea which md5 will be returned.
Sounds like a bad design.
Alternative idea: For each file, collect the md5 into a map `hash:filename`. Then check for the next file if its hash is already stored in the map, and if not add it to the map.

Reply:
In this function, file size will be checked first to speed up the checking process.
A simple `hash:filename` can not do that.
Should I remove the file size checking?

Reply:
improved in https://github.com/go-gitea/gitea/pull/24872/commits/3ef4aa814d52cb24c24ac7c760465258782731a0